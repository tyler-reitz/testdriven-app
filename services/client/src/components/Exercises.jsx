import React, { Component } from 'react'
import axios from 'axios'
import AceEditor from 'react-ace'
import brace from 'brace'
import 'brace/mode/python'
import 'brace/theme/solarized_dark'

class Exercises extends Component {
  constructor(props) {
    super(props)
    this.state = {
      exercises: [],
      editor: {
        value: '# Enter your code here.',
        button: { isDisabled: false, },
        showGrading: false,
        showCorrect: false,
        showIncorrect: false,
      },
      url: process.env.REACT_APP_API_GATEWAY_URL,
    }
  }

  componentDidMount() {
    this.getExercises()
    console.log(process.env.REACT_APP_API_GATEWAY_URL)
  }

  getExercises() {
    axios.get(`${process.env.REACT_APP_EXERCISES_SERVICE_URL}/exercises`)
      .then((res) => { this.setState({ exercises: res.data.data.exercises }) })
      .catch((err) => { console.log(err) })
  }

  onChange = (value) => {
    const newState = this.state.editor
    newState.value = value
    this.setState(newState)
  }

  submitExercise = (evt) => {
    evt.preventDefault()

    const newState = this.state.editor
    newState.showGrading = true
    newState.showCorrect = false
    newState.showIncorrect = false
    newState.button.isDisabled = true 

    this.setState(newState)

    const data = { answer: this.state.editor.value }
    const url = process.env.REACT_APP_API_GATEWAY_URL
    
    axios.post(url, data)
      .then((res) => {
        console.log(res) 
        newState.showGrading = false
        newState.button.isDisabled = false
        if (res.data) { newState.showCorrect = true }
        if (!res.data) { newState.showIncorrect = true }
        this.setState(newState)
      })
      .catch((err) => {
        newState.showGrading = false
        newState.button.isDisabled = false
        console.log(err)
      })
  }

  render() {
    return (
      <div>
        <h1 className="title is-1">Exercises</h1>
        <hr/><br/>
        {!this.props.isAuthenticated &&
          <div className="notification is-warning">
            <span>Please log in to submit an exercise.</span>
          </div>
        }
        {this.state.exercises.length &&
          <div key={this.state.exercises[0].id}>
            <h5 className="title is-5">{this.state.exercises[0].body}</h5>
            <AceEditor 
              mode="python"
              theme="solarized_dark"
              name={(this.state.exercises[0].id).toString()}
              onLoad={this.onLoad}
              onChange={this.onChange}
              fontSize={14}
              height={'175px'}
              showPrintMargin={true}
              showGutter={true}
              highlightActiveLine={true}
              value={this.state.editor.value}
              style={{
                marginBottom: '10px'
              }}
              editorProps={{
                $blockScrolling: Infinity
              }}
            />
            {this.props.isAuthenticated && 
              <div>
                <button 
                  className="button is-primary" 
                  onClick={this.submitExercise}
                  disabled={this.state.editor.button.isDisabled}
                >Run Code</button>
                {this.state.editor.showGrading &&
                  <h5 className="title is-5">
                    <span className="icon is-large">
                      <i className="fas fa-spinner fa-pulse"></i>
                    </span>
                    <span className="grade-text">Grading...</span>
                  </h5>
                }
                {this.state.editor.showCorrect &&
                  <h5 className="title is-5">
                    <span className="icon is-large">
                      <i className="fas fa-check"></i>
                    </span>
                    <span className="grade-text">Correct!</span>
                  </h5>
                }
                {this.state.editor.showIncorrect &&
                  <h5 className="title is-5">
                    <span className="icon is-large">
                      <i className="fas fa-times"></i>
                    </span>
                    <span className="grade-text">Incorrect!</span>
                  </h5>
                }
              </div>
            }
            <br/><hr/>
          </div>
        }
      </div>
    )
  }
}

export default Exercises
