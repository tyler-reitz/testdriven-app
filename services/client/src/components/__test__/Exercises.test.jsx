import React from 'react'
import { shallow, mount } from 'enzyme'
import renderer from 'react-test-renderer'

import AceEditor from 'react-ace'
jest.mock('react-ace')

import Exercises from '../Exercises'

it('Exercises renders properly when not authenticated', () => {
  const wrapper = shallow(<Exercises isAuthenticated={false} />)
  const element = wrapper.find('h5')
  expect(element.length).toBe(1)
  const alert = wrapper.find('.notification')
  expect(alert.length).toBe(1)
  const alertMessage = wrapper.find('.notification > span')
  expect(alertMessage.get(0).props.children).toContain(
    'Please log in to submit an exercise.'
  )
})

it('Exercises renders properly when authenticated', () => {
  const wrapper = shallow(<Exercises isAuthenticated={true} />)
  const heading = wrapper.find('h5')
  expect(heading.length).toBe(1)
  const alert = wrapper.find('.notifiction')
  expect(alert.length).toBe(0)
})

it('Exercises renders a snapshot properly', () => {
  const tree = renderer.create(<Exercises />).toJSON()
  expect(tree).toMatchSnapshot()
})

it('Exercises will call componentWillMount when mounted', () => {
  const onWillMount = jest.fn()
  Exercises.prototype.componentWillMount = onWillMount
  const wrapper = mount(<Exercises />)
  expect(onWillMount).toHaveBeenCalledTimes(1)
})
