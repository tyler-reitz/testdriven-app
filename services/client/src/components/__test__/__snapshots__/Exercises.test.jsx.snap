// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Exercises renders a snapshot properly 1`] = `
<div>
  <h1
    className="title is-1"
  >
    Exercises
  </h1>
  <hr />
  <br />
  <div
    className="notification is-warning"
  >
    <span>
      Please log in to submit an exercise.
    </span>
  </div>
  <div>
    <h5
      className="title is-5"
    >
      Define a function called sum that takes
      two integers as arguments and returns their sum.
    </h5>
    <br />
    <hr />
  </div>
</div>
`;
